component/kafka/group/component.go

```diff
@@ -0,0 +1,425 @@
+	}
+
+	defaultSaramaCfg, err := kafka.DefaultSaramaConfig(name)
+	if err != nil {
+		return nil, err
+	}
+
+	cmp := &Component{
+		name:         name,
+		group:        group,
+		brokers:      brokers,
+		topics:       topics,
+		proc:         proc,
+		retries:      defaultRetries,
+		retryWait:    defaultRetryWait,
+		batchSize:    defaultBatchSize,
+		batchTimeout: defaultBatchTimeout,
+		failStrategy: defaultFailureStrategy,
+		saramaConfig: defaultSaramaCfg,
+	}
+
+	for _, optionFunc := range oo {
+		err = optionFunc(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Component is a kafka consumer implementation that processes messages in batch
+type Component struct {
+	name         string
+	group        string
+	topics       []string
+	brokers      []string
+	saramaConfig *sarama.Config
+	proc         kafka.BatchProcessorFunc
+	failStrategy kafka.FailStrategy
+	batchSize    uint
+	batchTimeout time.Duration
+	retries      uint
+	retryWait    time.Duration
+	commitSync   bool
+}
+
+// Run starts the consumer processing loop to process messages from Kafka.
+func (c *Component) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	return c.processing(ctx)
+}
+
+func (c *Component) processing(ctx context.Context) error {
+	var componentError error
+
+	retries := int(c.retries)
+	for i := 0; i <= retries; i++ {
+		handler := newConsumerHandler(ctx, c.name, c.group, c.proc, c.failStrategy, c.batchSize,
+			c.batchTimeout, c.commitSync)
+
+		client, err := sarama.NewConsumerGroup(c.brokers, c.group, c.saramaConfig)
+		if err != nil {
+			componentError = err
+			log.Errorf("error creating consumer group client for kafka component: %v", err)
+		}
+
+		if client != nil {
+			// `Consume` should be called inside an infinite loop, when a
+			// server-side rebalance happens, the consumer session will need to be
+			// recreated to get the new claims
+			if err := client.Consume(ctx, c.topics, handler); err != nil {
+				log.Errorf("error from kafka consumer: %v", err)
+			}
+
+			// check if context was cancelled or deadline exceeded, signaling that the consumer should stop
+			if ctx.Err() != nil {
+				log.Infof("kafka component terminating: context cancelled or deadline exceeded")
+				break
+			}
+
+			err = client.Close()
+			if err != nil {
+				log.Errorf("error closing kafka consumer: %v", err)
+			}
+		}
+
+		consumerErrorsInc(c.name)
+		if c.retries > 0 {
+			if handler.processedMessages {
+				i = 0
+			}
+			log.Errorf("failed run, retry %d/%d with %v wait: %v", i, c.retries, c.retryWait, err)
+			time.Sleep(c.retryWait)
+		}
+
+		// If there is no component error which is a result of not being able to initialize the consumer
+		// then the handler errored while processing a message. This faulty message is then the reason
+		// behind the component failure.
+		if i == retries && componentError == nil {
+			componentError = handler.err
```

Comment:
I would actually wrap it e.g. `fmt.Errorf("handler failure exhausted %d retries: %w", i, handler.err)`

Reply:
What I am thinking is: how can the receiver of this error quickly distinguish between handler error and Kafka/client error?
But no need to make changes, we can improve it later

Reply:
The handler is an internal abstraction. Perhaps 
```
fmt.Errorf("potentially poisonous message failure exhausted %d retries: %w", i, handler.err)
```

Reply:
The processor func can fail for other reasons though (bugs in the code, last but not least); perhaps let's not mention poisonous messages there?

Reply:
```
fmt.Errorf("message processing failure exhausted %d retries: %w", i, handler.err)
```
perhaps?

Reply:
done 👍